platooning_tactical_plugin/src/platooning_tactical_plugin.cpp

```diff
@@ -0,0 +1,617 @@
+    ROS_DEBUG_STREAM("distance: " << distance);
+    ROS_DEBUG_STREAM("p: " << p.point.x() << ", " << p.point.y());
+    if (distance < min_distance)
+    {
+      best_index = i;
+      min_distance = distance;
+    }
+    i++;
+  }
+
+  return best_index;
+}
+
+
+void PlatooningTacticalPlugin::splitPointSpeedPairs(const std::vector<PointSpeedPair>& points,
+                                                std::vector<lanelet::BasicPoint2d>* basic_points,
+                                                std::vector<double>* speeds)
+{
+  basic_points->reserve(points.size());
+  speeds->reserve(points.size());
+
+  for (const auto& p : points)
+  {
+    basic_points->push_back(p.point);
+    speeds->push_back(p.speed);
+  }
+}
+
+std::unique_ptr<smoothing::SplineI>
+PlatooningTacticalPlugin::compute_fit(const std::vector<lanelet::BasicPoint2d>& basic_points)
+{
+  if (basic_points.size() < 3)
+  {
+    ROS_WARN_STREAM("Insufficient Spline Points");
+    return nullptr;
+  }
+
+  std::unique_ptr<smoothing::SplineI> spl = std::make_unique<smoothing::CubicSpline>();
+  spl->setPoints(basic_points);
+
+  return spl;
+}
+
+std::vector<cav_msgs::TrajectoryPlanPoint> PlatooningTacticalPlugin::compose_trajectory_from_centerline(
+    const std::vector<PointSpeedPair>& points, const cav_msgs::VehicleState& state)
+{
+  ROS_DEBUG_STREAM("VehicleState: "
+                   << " x: " << state.X_pos_global << " y: " << state.Y_pos_global << " yaw: " << state.orientation
+                   << " speed: " << state.longitudinal_vel);
+
+  ROS_DEBUG_STREAM("points size: " << points.size());
+  
+  log::printDebugPerLine(points, &log::pointSpeedPairToStream);
+
+  int nearest_pt_index = getNearestPointIndex(points, state);
+
+  ROS_DEBUG_STREAM("NearestPtIndex: " << nearest_pt_index);
+
+  std::vector<PointSpeedPair> future_points(points.begin() + nearest_pt_index + 1, points.end()); // Points in front of current vehicle position
+
+  auto time_bound_points = constrain_to_time_boundary(future_points, config_.trajectory_time_length);
+
+  ROS_DEBUG_STREAM("time_bound_points: " << time_bound_points.size());
+
+  log::printDebugPerLine(time_bound_points, &log::pointSpeedPairToStream);
+
+  ROS_DEBUG("Got basic points ");
+  std::vector<DiscreteCurve> sub_curves = compute_sub_curves(time_bound_points);
+
+  ROS_DEBUG_STREAM("Got sub_curves " << sub_curves.size());
+
+  std::vector<double> final_yaw_values;
+  std::vector<double> final_actual_speeds;
+  std::vector<lanelet::BasicPoint2d> all_sampling_points;
+
+  for (const auto& discreet_curve : sub_curves)
+  {
+    ROS_DEBUG("SubCurve");
+
+    std::vector<double> speed_limits;
+    std::vector<lanelet::BasicPoint2d> curve_points;
+    splitPointSpeedPairs(discreet_curve.points, &curve_points, &speed_limits);
+
+    std::unique_ptr<smoothing::SplineI> fit_curve = compute_fit(curve_points); // Compute splines based on curve points
+
+    if (!fit_curve)
+    {  // TODO how better to handle this case
```

Comment:
Is this todo from the original ILC-plugin, or an additional thing we should be tracking with respect to platooning plugin?

Reply:
it is from ILC. I dont think that would be an issue in platooning, since platooning does not work well on highly curved roads, and probably not activated on that type of environment.